File: app/src/main/java/com/rowland/delivery/features/dash/presentation/fragments/NewProductFragment.java
package com.rowland.delivery.features.dash.presentation.fragments;


import android.Manifest;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.firebase.auth.FirebaseAuth;
import com.rowland.delivery.features.dash.di.modules.product.ProductModule;
import com.rowland.delivery.features.dash.domain.models.product.Product;
import com.rowland.delivery.features.dash.presentation.activities.DashActivity;
import com.rowland.delivery.features.dash.presentation.viewmodels.product.NewProductViewModel;
import com.rowland.delivery.merchant.R;
import com.tbruyelle.rxpermissions2.RxPermissions;

import java.util.Date;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

import static android.support.v4.app.FragmentManager.POP_BACK_STACK_INCLUSIVE;


public class NewProductFragment extends Fragment {

    private Unbinder unbinder;
    private NewProductViewModel newProductViewModel;

    @BindView(R.id.main_toolbar_newproduct)
    Toolbar toolbar;

    @BindView(R.id.imageview)
    ImageView productImage;

    @BindView(R.id.input_meal_title)
    EditText edtTitle;

    @BindView(R.id.input_meal_description)
    EditText edtDescription;

    @BindView(R.id.input_meal_quantity)
    EditText edtQuantity;

    @BindView(R.id.input_meal_unit)
    Spinner edtUnit;

    @BindView(R.id.input_meal_price)
    EditText edtPrice;

    @BindView(R.id.input_meal_stock)
    EditText edtStock;

    @Inject
    @Named("newproduct")
    ViewModelProvider.Factory newProductFactory;


    public NewProductFragment() {
        // Required empty public constructor
    }

    public static Fragment newInstance(Bundle args) {
        NewProductFragment frag = new NewProductFragment();
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        newProductViewModel = ViewModelProviders.of(this, newProductFactory).get(NewProductViewModel.class);
        newProductViewModel.setFirebaseUserUid(FirebaseAuth.getInstance().getCurrentUser().getUid());

        if (getArguments() != null) {
            String category = getArguments().getString("selected_category");
            newProductViewModel.setCategory(category);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_new_product, container, false);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        ((DashActivity) getActivity()).getDashComponent()
                .getProductComponent(new ProductModule())
                .inject(this);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);

        if (toolbar != null) {
            ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
        }

        newProductViewModel.getImages().observe(this, uris -> {
            Glide.with(this)
                    .load(uris.get(0))
                    .crossFade()
                    .fitCenter()
                    .diskCacheStrategy(DiskCacheStrategy.NONE)
                    .skipMemoryCache(true)
                    .into(productImage);
        });
    }

    @OnClick(R.id.fab_addimage)
    public void addImage() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            new RxPermissions(getActivity()).request(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    .subscribe(granted -> {
                        if (!granted) {
                            Toast.makeText(getActivity(), "Please grant permissions to select image", Toast.LENGTH_SHORT).show();
                        } else {
                            newProductViewModel.selectImages();
                        }
                    });
        }
    }

    @OnClick(R.id.btn_save)
    public void saveProduct() {

        Product product = new Product();
        product.setPrice(Integer.valueOf(edtPrice.getText().toString()));
        product.setName(edtTitle.getText().toString());
        product.setSaleQuantity(Integer.valueOf(edtQuantity.getText().toString()));
        product.setItemQuantity(Integer.valueOf(edtStock.getText().toString()));
        product.setDescription(edtDescription.getText().toString());
        product.setCreatedAt(new Date());

        newProductViewModel.saveProduct(product)
                .subscribe(newProduct -> {
                    Toast.makeText(getActivity(), "Product added succesfully", Toast.LENGTH_SHORT).show();
                    getActivity().getSupportFragmentManager().popBackStack(NewProductFragment.class.getSimpleName(), POP_BACK_STACK_INCLUSIVE);
                }, throwable -> {
                    Toast.makeText(getActivity(), "Product not added", Toast.LENGTH_SHORT).show();
                });
    }

    @OnClick(R.id.btn_cancell)
    public void cancell() {
        getActivity().getSupportFragmentManager().popBackStack(NewProductFragment.class.getSimpleName(), POP_BACK_STACK_INCLUSIVE);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        unbinder.unbind();
    }
}

File: app/src/main/java/com/rowland/delivery/features/dash/presentation/fragments/NewProductFragment.kt
package com.rowland.delivery.features.dash.presentation.fragments


import android.Manifest
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager.POP_BACK_STACK_INCLUSIVE
import androidx.appcompat.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.meetic.shuffle.Shuffle
import com.meetic.shuffle.ShuffleViewAnimatorOnSecondCard
import com.rowland.delivery.features.dash.di.modules.product.ProductModule
import com.rowland.delivery.features.dash.domain.models.product.ProductEntity
import com.rowland.delivery.features.dash.presentation.activities.DashActivity
import com.rowland.delivery.features.dash.presentation.viewmodels.product.NewProductViewModel
import com.rowland.delivery.merchant.R
import com.tbruyelle.rxpermissions2.RxPermissions
import kotlinx.android.synthetic.main.content_product_images_shuffle.view.*
import kotlinx.android.synthetic.main.fragment_new_product.*
import java.util.*
import javax.inject.Inject


class NewProductFragment : Fragment() {

    private lateinit var newProductViewModel: NewProductViewModel

    @Inject
    lateinit var newProductFactory: ViewModelProvider.Factory

    companion object {
        fun newInstance(args: Bundle?): androidx.fragment.app.Fragment {
            val frag = NewProductFragment()
            frag.arguments = args
            return frag
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        newProductViewModel = ViewModelProviders.of(this, newProductFactory).get(NewProductViewModel::class.java)

        if (arguments != null) {
            val category = arguments!!.getString("selected_category")
            newProductViewModel.setCategory(category!!)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_new_product, container, false)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        (activity as DashActivity).dashComponent.getProductComponent(ProductModule()).inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        (activity as AppCompatActivity).setSupportActionBar(newproduct_toolbar)

        fab_addimage.setOnClickListener { view ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                RxPermissions(activity!!).request(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                        .subscribe { granted ->
                            if (!granted) {
                                Toast.makeText(activity, "Please grant permissions to select image", Toast.LENGTH_SHORT).show()
                            } else {
                                newProductViewModel.selectImages()
                            }
                        }
            }
        }

        btn_save.setOnClickListener { view ->
            val product = ProductEntity()
            product.price = Integer.valueOf(input_product_pricing.text!!.toString())
            product.name = input_product_name.text!!.toString()
            product.saleQuantity = Integer.valueOf(input_product_quantity.text!!.toString())
            product.itemQuantity = Integer.valueOf(input_product_stock!!.text!!.toString())
            product.description = input_product_description.text!!.toString()
            product.createdAt = Date()

            newProductViewModel.saveProduct(product)
                    .subscribe({ newProduct ->
                        Toast.makeText(activity, "ProductEntity added succesfully", Toast.LENGTH_SHORT).show()
                        activity!!.supportFragmentManager.popBackStack(NewProductFragment::class.java.simpleName, POP_BACK_STACK_INCLUSIVE)
                    }) { throwable -> Toast.makeText(activity, "ProductEntity not added", Toast.LENGTH_SHORT).show() }

        }

        btn_cancell.setOnClickListener { view ->
            activity!!.supportFragmentManager.popBackStack(NewProductFragment::class.java.simpleName, POP_BACK_STACK_INCLUSIVE)
        }

        newProductViewModel.images.observe(this, Observer { uris ->
            new_product_shuffle.shuffleSettings.numberOfDisplayedCards = uris!!.size
            new_product_shuffle.shuffleAdapter = ImageShuffleAdapter(uris)
            new_product_shuffle.viewAnimator = ShuffleViewAnimatorOnSecondCard()
        })

    }

    class ImageShuffleViewHolder(itemView: View) : Shuffle.ViewHolder(itemView) {
        fun bind(uri: Uri, position: Int) {
            itemView!!.product_images_count!!.text = position.toString()
            val options = RequestOptions()
            options.centerCrop().diskCacheStrategy(DiskCacheStrategy.ALL)
            Glide.with(itemView!!.product_image.context)
                    .load(uri)
                    .apply(options)
                    .into(itemView!!.product_image)
        }
    }

    class ImageShuffleAdapter(_uris: List<Uri>) : Shuffle.Adapter<ImageShuffleViewHolder>() {
        var uris: List<Uri>

        init {
            uris = _uris
        }

        override fun onCreateViewHolder(parent: ViewGroup, type: Int): ImageShuffleViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.content_product_images_shuffle, parent, false)
            return ImageShuffleViewHolder(view)
        }

        override fun onBindViewHolder(viewHolder: ImageShuffleViewHolder, position: Int) {
            viewHolder.bind(this.uris.get(position), position)
        }

        override fun getItemCount(): Int {
            return uris.size
        }
    }
}
